webroot/src/store/users/users.spec.ts

```diff
@@ -0,0 +1,440 @@
+    });
+    it('should successfully update user (not already in store - empty store)', () => {
+        const state = {};
+        const user = { id: 1 };
+
+        mutations[MutationTypes.STORE_UPDATE_USER](state, user);
+
+        expect(state.model).to.matchPattern([user]);
+    });
+    it('should successfully update user (not already in store - non-empty store)', () => {
+        const state = { model: [{ id: 2 }]};
+        const user = { id: 1 };
+
+        mutations[MutationTypes.STORE_UPDATE_USER](state, user);
+
+        expect(state.model).to.matchPattern([ { id: 2 }, user]);
+    });
+    it('should successfully update user (already in store)', () => {
+        const state = { model: [{ id: 1, name: 'test1' }]};
+        const user = { id: 1, name: 'test2' };
+
+        mutations[MutationTypes.STORE_UPDATE_USER](state, user);
+
+        expect(state.model).to.matchPattern([user]);
+    });
+    it('should successfully remove user (id missing)', () => {
+        const state = {};
+        const userId = '';
+
+        mutations[MutationTypes.STORE_REMOVE_USER](state, userId);
+
+        expect(state).to.equal(state);
+    });
+    it('should successfully remove user (not already in store)', () => {
+        const state = {};
+        const userId = 1;
+
+        mutations[MutationTypes.STORE_REMOVE_USER](state, userId);
+
+        expect(state).to.equal(state);
+    });
+    it('should successfully remove user (already in store)', () => {
+        const state = { model: [{ id: 1 }, { id: 2 }]};
+        const userId = 1;
+
+        mutations[MutationTypes.STORE_REMOVE_USER](state, userId);
+
+        expect(state.model).to.matchPattern([{ id: 2 }]);
+    });
+    it('should successfully remove user (already in store - only record)', () => {
+        const state = { model: [{ id: 1 }]};
+        const userId = 1;
+
+        mutations[MutationTypes.STORE_REMOVE_USER](state, userId);
+
+        expect(state.model).to.matchPattern([]);
+    });
+    it('should successfully reset users store', () => {
+        const state = {
+            model: [{ id: 1 }],
+            total: 1,
+            isLoading: true,
+            error: new Error(),
+        };
+
+        mutations[MutationTypes.STORE_RESET_USERS](state);
+
+        expect(state).to.matchPattern({
+            model: null,
+            total: 0,
+            isLoading: false,
+            error: null,
+        });
+    });
+});
+describe('Users Store Actions', async () => {
+    it('should successfully start get-users request with next page', async () => {
+        const commit = sinon.spy();
+        const dispatch = sinon.spy();
+        const params = { getNextPage: true };
+
+        await actions.getUsersRequest({ commit, getters, dispatch }, { params });
+
+        expect(commit.calledOnce).to.equal(true);
+        expect(commit.firstCall.args).to.matchPattern([MutationTypes.GET_USERS_REQUEST]);
+        expect(dispatch.callCount).to.equal(4);
```

Comment:
More for my understanding: What's the justification for testing the number of dispatches? Isn't this purely a function of how you structure your vuex actions? It doesn't obviously seem like something worth testing, but I imagine I'm missing something.

Reply:
We used to not include it. Then we ran into cases where we'd accidentally leave in dispatches to other actions for temp stuff / debugging / old-flows / whatever. Not often, but when it happened, this simple test case would have caught it. So for some of the more robust actions, we've gotten used to adding it. Mainly as a protection against unnecessary action dispatches.

Reply:
Got it- that makes sense.